File: crates/encore_ecl/src/position.rs
use std::fmt;

/// A location in an ECL source file.
/// Lines and columns are 1-based; columns count runes, not bytes.
#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
pub struct Position {
    pub file: String,
    /// byte offset within the file
    pub offset: usize,
    pub line: usize,
    pub column: usize,
}

impl Position {
    /// Reports whether the position refers to an actual source location.
    pub fn is_valid(&self) -> bool {
        self.line > 0
    }
}

impl fmt::Display for Position {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match (self.file.is_empty(), self.is_valid()) {
            (true, false) => write!(f, "<unknown position>"),
            (_, false) => write!(f, "{}", self.file),
            (true, _) => write!(f, "{}:{}", self.line, self.column),
            _ => write!(f, "{}:{}:{}", self.file, self.line, self.column),
        }
    }
}

/// A contiguous range of source text.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Span {
    pub start: Position,
    pub end: Position,
}

/// Holds the contents of a parsed file for snippet rendering.
#[derive(Debug)]
pub(crate) struct SourceFile {
    pub(crate) name: String,
    pub(crate) src: String,
    /// byte offset of the start of each line, 0-based index
    line_start: Vec<usize>,
}

impl SourceFile {
    pub(crate) fn new(name: impl Into<String>, src: impl Into<String>) -> SourceFile {
        let src = src.into();
        let mut starts = vec![0usize];
        for (i, b) in src.bytes().enumerate() {
            if b == b'\n' {
                starts.push(i + 1);
            }
        }
        SourceFile {
            name: name.into(),
            src,
            line_start: starts,
        }
    }

    /// Returns the text of the given 1-based line, without the trailing newline.
    pub(crate) fn line(&self, n: usize) -> Option<&str> {
        if n < 1 || n > self.line_start.len() {
            return None;
        }
        let start = self.line_start[n - 1];
        let end = if n < self.line_start.len() {
            self.line_start[n] - 1
        } else {
            self.src.len()
        };
        Some(self.src[start..end].trim_end_matches('\r'))
    }
}

File: pkg/ecl/position.go
package ecl

import (
	"fmt"
	"strings"
)

// Position is a location in an ECL source file.
// Lines and columns are 1-based; columns count runes, not bytes.
type Position struct {
	File   string
	Offset int // byte offset within the file
	Line   int
	Column int
}

// IsValid reports whether the position refers to an actual source location.
func (p Position) IsValid() bool { return p.Line > 0 }

func (p Position) String() string {
	switch {
	case p.File == "" && !p.IsValid():
		return "<unknown position>"
	case !p.IsValid():
		return p.File
	case p.File == "":
		return fmt.Sprintf("%d:%d", p.Line, p.Column)
	default:
		return fmt.Sprintf("%s:%d:%d", p.File, p.Line, p.Column)
	}
}

// Span is a contiguous range of source text.
type Span struct {
	Start, End Position
}

// sourceFile holds the contents of a parsed file for snippet rendering.
type sourceFile struct {
	name      string
	src       string
	lineStart []int // byte offset of the start of each line, 0-based index
}

func newSourceFile(name, src string) *sourceFile {
	starts := []int{0}
	for i := 0; i < len(src); i++ {
		if src[i] == '\n' {
			starts = append(starts, i+1)
		}
	}
	return &sourceFile{name: name, src: src, lineStart: starts}
}

// line returns the text of the given 1-based line, without the trailing newline.
func (f *sourceFile) line(n int) (string, bool) {
	if f == nil || n < 1 || n > len(f.lineStart) {
		return "", false
	}
	start := f.lineStart[n-1]
	end := len(f.src)
	if n < len(f.lineStart) {
		end = f.lineStart[n] - 1
	}
	return strings.TrimSuffix(f.src[start:end], "\r"), true
}
